### ISSUE TITLE
Odd and unexpected compiler error

### ISSUE THREAD

**Issue by bananabrick** [rust-lang/rust#95381]
This code won't compile 
I tried this code:

```rust
#[derive(Debug)]
struct Node {
    value: i32,
    parent: RefCell<Weak<Node>>,
    children: RefCell<Vec<Rc<Node>>>,
}

fn main() {
    let leaf = Rc::new(Node {
        value: 3,
        parent: RefCell::new(Weak::new()),
        children: RefCell::new(vec![]),
    });

    println!("leaf parent = {:?}", leaf.parent.borrow().upgrade());

    let branch = Rc::new(Node {
        value: 5,
        parent: RefCell::new(Weak::new()),
        children: RefCell::new(vec![Rc::clone(&leaf)]),
    });

    *leaf.parent.borrow_mut() = Rc::downgrade(&branch);

    if let Some(x) = leaf.parent.borrow().upgrade() {
        println!("leaf parent = {:?}", x);
    }
}
```

I expected to see this happen: code compiled

Instead, this happened: compiler error
```
a temporary with access to the borrow is created here ...rustcE0597
main.rs(35, 22): original diagnostic
`leaf` does not live long enough
borrowed value does not live long enoughrustcE0597
main.rs(38, 1): `leaf` dropped here while still borrowed
main.rs(35, 22): a temporary with access to the borrow is created here ...
main.rs(38, 1): ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `Ref<'_, std::rc::Weak<Node>>`
main.rs(37, 6): consider adding semicolon after the expression so its temporaries are dropped sooner, before the local variables declared by the block are dropped: `;`
```

**Comment by scottmcm** [comment]
Working playground: <https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ef9fc1e647fb3d4c24bc023338d24f44>

Real error, rather than the IDE-neutered one:
```
error[[E0597]](https://doc.rust-lang.org/nightly/error-index.html#E0597): `leaf` does not live long enough
  [--> src/main.rs:28:22
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021#)   |
28 |     if let Some(x) = leaf.parent.borrow().upgrade() {
   |                      ^^^^----------------
   |                      |
   |                      borrowed value does not live long enough
   |                      a temporary with access to the borrow is created here ...
...
31 | }
   | -
   | |
   | `leaf` dropped here while still borrowed
   | ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `Ref<'_, std::rc::Weak<Node>>`
   |
help: consider adding semicolon after the expression so its temporaries are dropped sooner, before the local variables declared by the block are dropped
   |
30 |     };
   |      +
```

Note that this compiles if one adds the `;` as suggested by the `help`.

**Comment by bananabrick** [comment]
Thanks, @scottmcm.

But I don't understand what's going on. Why is this error happening in the first place?

**Comment by bananabrick** [comment]
This code also doesn't compile for the same reasons.

```
use std::cell::RefCell;
use std::rc::{Rc, Weak};

#[derive(Debug)]
struct Node {
    value: i32,
    parent: RefCell<Weak<Node>>,
    children: RefCell<Vec<Rc<Node>>>,
}

fn main() {
    let leaf = Rc::new(Node {
        value: 3,
        parent: RefCell::new(Weak::new()),
        children: RefCell::new(vec![]),
    });

    println!("leaf parent = {:?}", leaf.parent.borrow().upgrade());

    let branch = Rc::new(Node {
        value: 5,
        parent: RefCell::new(Weak::new()),
        children: RefCell::new(vec![Rc::clone(&leaf)]),
    });

    *leaf.parent.borrow_mut() = Rc::downgrade(&branch);

    let x = leaf.parent.borrow().upgrade();
    if let Some(_) = leaf.parent.borrow().upgrade() {
        // println!("leaf parent = {:?}", x);
    }
}

```

No local variables defined in the block.

**Comment by zirconium-n** [comment]
Minimized version

``` rust
use std::cell::RefCell;
use std::rc::Weak;

fn main() {
    let leaf = RefCell::new(Weak::<()>::new());

    if let Some(_) = leaf.borrow().upgrade() {}
}
```

**Comment by cyrgani** [comment]
This is fixed in the 2024 edition by shortening the temporary's scope to be dropped sooner: [edition guide](https://doc.rust-lang.org/nightly/edition-guide/rust-2024/temporary-tail-expr-scope.html)
